### PR TITLE
Changepoints that can cope with non-Krun data

### PR DESCRIPTION
**Pull request by snim2**

### CODE REVIEW

File: bin/mark_changepoints_in_json [hunk 1]
```diff
@@ -17,7 +18,108 @@ sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(
...
+            return False
+        return True
+
+    @property
+    def n(self):
+        return self.end - self.start
+
+class Segments(object):
+    """A list of Segments for a whole run sequence.
+    """
```
Comment (ltratt) [inline bin/mark_changepoints_in_json:50]: Missing blank line?

File: bin/mark_changepoints_in_json [hunk 2]
```diff
@@ -17,7 +18,108 @@ sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(
...
+            classification = 'flat'
+        # Slowdown -> last mean+ci does not overlap with another mean+ci and
+        #             it is higher than the first mean+ci.
+        elif self.segments[-1].mean == max(self.segments, key=lambda s: s.mean).mean:
+            classification = 'slowdown'
+        # Warmup -> last mean+ci does not overlap with another segment CI and
+        #           is lower than the first mean+ci.
+        elif (self.segments[-1].mean == min(self.segments, key=lambda s: s.mean).mean and
+              self.segments[0].mean == max(self.segments, key=lambda s: s.mean).mean):
+            classification = 'warm-up'
```
Comment (ltratt) [inline bin/mark_changepoints_in_json]: I think we're going for "warmup".

File: bin/mark_changepoints_in_json [hunk 3]
```diff
@@ -17,7 +18,108 @@ sys.path.insert(0, os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(
...
+        # Slow down -> not flat or a warm-up, and no changepoints after we
+        #              expect a steady state.
+        elif self.segments[-1].start < (self.length - self.steady_state):
+            classification = 'slowdown'
+        # No steady state -> there are changes after we expect steady state.
+        elif self.segments[-1].start > (self.length - self.steady_state):
+            classification = 'no steady state'
+        else:
+            classification = 'could not classify'
```
Comment (ltratt) [inline bin/mark_changepoints_in_json]: This should be impossible I think? If so, should it raise an exception?

File: bin/mark_changepoints_in_json [hunk 4]
```diff
@@ -46,49 +148,59 @@ def main(in_files, window_size):
...
                 else:
                     outliers = list()
-                c_points_all, means = get_changepoints(cpt, p_exec[:], outliers)
-                # The last datum is always a changepoint, so we ignore it.
-                c_points = c_points_all[:-1]
-                changepoints[bench].append(c_points)
-                changepoint_means[bench].append(means)
-                classifications[bench].append(get_classification(p_exec, c_points,
-                                                                 means, window_size))
+                segments = get_segments(cpt, p_exec[:], steady_state, outliers,
```
Comment (ltratt) [inline bin/mark_changepoints_in_json]: Does `get_segments` really mutate its second argument? I think it preemptively copies it, which means we can get rid of `[:]`.

File: bin/mark_changepoints_in_json [hunk 5]
```diff
@@ -46,49 +148,59 @@ def main(in_files, window_size):
...
         del p_exec[index]
     measurements = rpy2.robjects.FloatVector(p_exec)
-    changepoints = cpt.cpt_meanvar(measurements, method='PELT')
+    if use_snr_penalty:  # User has asked for SNR penalty.
+        try:  # Check it doesn't return NaN or Inf.
+            penalty = _get_snr_penalty(p_exec[-steady_state:])
+            changepoints = cpt.cpt_meanvar(measurements, method='PELT',
+                                           penalty='Manual', pen_value=penalty)
+        except ValueError:  # Use default penalty, 'MBIC'.
+            changepoints = cpt.cpt_meanvar(measurements, method='PELT')
```
Comment (ltratt) [inline bin/mark_changepoints_in_json:191]: Is PELT the same as MBIC?

Comment (snim2) [reply bin/mark_changepoints_in_json:191]: No, 'MBIC' would be an argument to 'penalty'. 'PELT' means we want 1 _or more_ changepoints.

Comment (ltratt) [reply bin/mark_changepoints_in_json:191]: OK, maybe the comment can be made a little clearer? I was confused by this.

Comment (ltratt) [inline bin/mark_changepoints_in_json]: OK, maybe the comment can be made clearer? I was confused by this.

Comment (ltratt) [reply bin/mark_changepoints_in_json:191]: OK, maybe the comment can be made clearer? I was confused by this.

**Comment by snim2** [comment]
OK, I think that answers everything. Squash?

**Comment by ltratt** [comment]
Yep, squash away!

**Comment by snim2** [comment]
Squashed.

**Review by ltratt — APPROVED** [review]

**Pull request merged**